internal/ingest/linksystem.go

```diff
@@ -182,82 +182,76 @@ func (ing *Ingester) ingestEntryChunk(publisher peer.ID, adCid cid.Cid, entryChu
 	return nil
 }
 
-// ingestAd fetches all the entries for a single advertisement
-func (ing *Ingester) ingestAd(publisher peer.ID, adCid cid.Cid) error {
+type adIngestError struct {
+	state adIngestState
+	err   error
+}
+
+type adIngestState string
+
+const (
+	adIngestDatastoreErr        adIngestState = "datastoreErr"
+	adIngestIndexerErr          adIngestState = "indexerErr"
+	adIngestDecodingErr         adIngestState = "decodingErr"
+	adIngestMalformedErr        adIngestState = "malformedErr"
+	adIngestRegisterProviderErr adIngestState = "registerErr"
+	adIngestSyncEntriesErr      adIngestState = "syncEntriesErr"
+	// Happens if there is an error during ingest of an entry chunk (rather than fetching it).
+	adIngestEntryChunkErr adIngestState = "ingestEntryChunkErr"
+)
```

Comment:
Does the code really care about these individual states, or only whether or not they are permanent type errors?  If the code does not care, then the description should be part of the error message, and use a `permanent bool` for the code to handle permanent and non-permanent errors differently.

Reply:
Let's continue this discussion here: https://github.com/filecoin-project/storetheindex/pull/237#discussion_r814924082